virtualmachine/aws/vm.go

```diff
@@ -358,6 +442,321 @@ func (vm *VM) Start() error {
+	_, err = svc.AuthorizeSecurityGroupEgress(egressInput)
+	if err != nil {
+		return fmt.Errorf("Failed to authorize security group egress rules: %v", err)
+	}
+
+	return nil
+}
+
+// RevokeSecurityGroup: Removes one or more rules from a security group
+func (vm *VM) RevokeSecurityGroup() error {
+	svc, err := getService(vm.Region)
+	if err != nil {
+		return fmt.Errorf("Failed to get AWS service: %v", err)
+	}
+
+	secGrp := vm.SecurityGroups[0]
+
+	ec2IpPermissions := ToEc2IpPermissions(secGrp.IpPermissions)
+	input := &ec2.RevokeSecurityGroupIngressInput{
+		GroupId:       &secGrp.Id,
+		IpPermissions: ec2IpPermissions}
+	_, err = svc.RevokeSecurityGroupIngress(input)
+	if err != nil {
+		return fmt.Errorf("Failed to revoke security group ingress rules: %v", err)
+	}
+
+	ec2IpPermissionsEgress := ToEc2IpPermissions(
+		secGrp.IpPermissionsEgress)
+	egressInput := &ec2.RevokeSecurityGroupEgressInput{
+		GroupId:       &secGrp.Id,
+		IpPermissions: ec2IpPermissionsEgress}
+	_, err = svc.RevokeSecurityGroupEgress(egressInput)
+	if err != nil {
+		return fmt.Errorf("Failed to revoke security group egress rules: %v", err)
+	}
+
+	return nil
+}
+
+// CreateVolume: Creates a volume with given parameter
+func (vm *VM) CreateVolume() error {
+	svc, err := getService(vm.Region)
+	if err != nil {
+		return fmt.Errorf("Failed to get AWS service: %v", err)
+	}
+
+	volume := vm.Volumes[0]
+	instanceStatus, err := GetInstanceStatus(svc, vm.InstanceID)
+	if err != nil {
+		return fmt.Errorf("Failed to get availability zone of instance: %v", err)
+	}
+	volume.AvailabilityZone = instanceStatus.AvailabilityZone
+
+	input := GetVolumeInput(&volume)
+	response, err := svc.CreateVolume(input)
+	if err != nil {
+		return fmt.Errorf("Failed to create volume: %v", err)
+	}
+
+	if err := waitUntilVolumeReady(svc, *response.VolumeId); err != nil {
+		return err
+	}
+
+	volume.VolumeId = *response.VolumeId
+	vm.Volumes[0] = volume
+
+	return nil
+}
+
+// AttachVolume: Attaches given volume to given instance
+func (vm *VM) AttachVolume() error {
+	svc, err := getService(vm.Region)
+	if err != nil {
+		return fmt.Errorf("Failed to get AWS service: %v", err)
+	}
+
+	volume := vm.Volumes[0]
+	input := &ec2.AttachVolumeInput{
+		Device:     &volume.DeviceName,
+		InstanceId: &vm.InstanceID,
+		VolumeId:   &volume.VolumeId}
+	_, err = svc.AttachVolume(input)
+	if err != nil {
+		return fmt.Errorf("Failed to attach volume: %v", err)
+	}
+
+	return nil
+}
+
+// DetachVolume: Detaches volume with given Id from instance
+func (vm *VM) DetachVolume() error {
+	svc, err := getService(vm.Region)
+	if err != nil {
+		return fmt.Errorf("Failed to get AWS service: %v", err)
+	}
+
+	volume := vm.Volumes[0]
+	input := &ec2.DetachVolumeInput{
+		VolumeId: &volume.VolumeId}
+	_, err = svc.DetachVolume(input)
+	if err != nil {
+		return fmt.Errorf("Failed to detach volume: %v", err)
+	}
+
+	if err := waitUntilVolumeReady(svc, volume.VolumeId); err != nil {
```

Comment:
do we need to wait till ready or something else?

Reply:
Actually function name should have been waitUntilVolumeAvailable, because stable state for volume after creation and when not being attached/used by VM is Available. But now have made the wait function for volume generic to match any given state.